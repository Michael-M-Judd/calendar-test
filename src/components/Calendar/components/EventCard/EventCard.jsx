import React from "react";
import styled, { css } from "styled-components";
import { getEventHeight } from "../../calendar.utils";

const cardBorderWidth = 5;

const cornerCardCss = css`
  border-top: ${cardBorderWidth * 0.75}px solid transparent;
  border-bottom: ${cardBorderWidth * 0.75}px solid transparent;
  border-left: ${cardBorderWidth * 1.1}px solid ${({ theme }) => theme.primary};
  border-radius: ${cardBorderWidth}px;
  position: absolute;
  height: 0px;
  width: 0px;
  content: "";
`;

const StyledCard = styled.div`
  background: ${({ theme }) => theme.background.white};
  border-left: ${cardBorderWidth}px ${({ theme }) => theme.primary} solid;
  border-radius: 0 2px 2px 0;
  display: flex;
  flex-direction: column;
  height: ${props => props.height * 0.8}px;
  padding: 10px;
  position: relative;
  margin: 0 5px;

  &:after {
    ${cornerCardCss}
    bottom: -${cardBorderWidth * 0.75}px;
    left: -${cardBorderWidth}px;
  }

  &:before {
    ${cornerCardCss}
    top: -${cardBorderWidth * 0.75}px;
    left: -${cardBorderWidth}px;
  }
`;

const Title = styled.span`
  color: ${({ theme }) => theme.primary};
  font-weight: bold;
  font-size: 18px;
  margin-bottom: 5px;
`;

const Location = styled.span`
  color: ${({ theme }) => theme.font.light};
  font-size: 12px;
`;

export const testKey = ({ start, end }) => `event-card:${start}:${end}`;

export const EventCard = ({ title, location, event }) => {
  return (
    <StyledCard height={getEventHeight(event)} data-testid={testKey(event)}>
      <Title>{title}</Title>
      <Location>{location}</Location>
    </StyledCard>
  );
};
