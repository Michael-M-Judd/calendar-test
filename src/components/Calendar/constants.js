// Height of a single hour
export const calendarHeight = 720;

export const periods = Object.freeze({
  am: "AM",
  pm: "PM"
});

export const hours = [
  { time: "9:00", period: periods.am },
  { time: "10:00", period: periods.am },
  { time: "11:00", period: periods.am },
  { time: "12:00", period: periods.pm },
  { time: "1:00", period: periods.pm },
  { time: "2:00", period: periods.pm },
  { time: "3:00", period: periods.pm },
  { time: "4:00", period: periods.pm },
  { time: "5:00", period: periods.pm },
  { time: "6:00", period: periods.pm },
  { time: "7:00", period: periods.pm },
  { time: "8:00", period: periods.pm },
  { time: "9:00", period: periods.pm }
];

export const hourBlockHeight = calendarHeight / (hours.length - 1);
