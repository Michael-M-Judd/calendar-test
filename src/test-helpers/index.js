import { render as _render } from "@testing-library/react";
import { Provider } from "../components/Provider/Provider";

export const render = (component, options) =>
  _render(component, { ...options, wrapper: Provider });
