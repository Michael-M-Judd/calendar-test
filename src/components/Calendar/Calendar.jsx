import React, { useMemo, useState } from "react";
import styled from "styled-components";
import { EventsContainer } from "./components/EventsContainer";
import { Timeline } from "./components/Timeline";
import { EventCard } from "./components/EventCard";
import { EventPosition } from "./components/EventPosition";
import { getOverlappingEvents } from "./calendar.utils";

const StyledDiv = styled.div`
  display: flex;
  margin: 50px;
`;

const getEventKey = ({ start, end }) => `${start}:${end}`;

export const Calendar = ({ initialEvents = [] }) => {
  const [events, setEvents] = useState(initialEvents);

  // For the purposes of testing - setting it on the window
  window.layoutDay = setEvents;

  const eventGroups = useMemo(() => getOverlappingEvents(events), [events]);

  return (
    <StyledDiv>
      <Timeline />
      <EventsContainer>
        {eventGroups.map(eventGroup =>
          eventGroup.map((event, index) => (
            <EventPosition
              event={event}
              previousColumnEvent={eventGroup[index - 2]}
              key={getEventKey(event)}
              order={index}
              overlapping={eventGroup.length > 1}
            >
              <EventCard
                event={event}
                title="Sample Item"
                location="Sample Location"
              />
            </EventPosition>
          ))
        )}
      </EventsContainer>
    </StyledDiv>
  );
};
