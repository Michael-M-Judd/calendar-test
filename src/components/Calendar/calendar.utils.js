import { hourBlockHeight } from "./constants";

const hourLength = 60;

/**
 * Gets the height in px of an event
 */
export const getEventHeight = ({ start, end }) => {
  return ((end - start) / hourLength) * hourBlockHeight;
};

/**
 * Finds the absolute y-position for a calendar event.
 */
export const getEventYPosition = ({ event, previousColumnEvent }) => {
  const baseEventPosition = (event.start / hourLength) * hourBlockHeight;

  if (
    // last event in column overlaps with latest event
    previousColumnEvent &&
    (previousColumnEvent.start <= event.start &&
      previousColumnEvent.end >= event.start)
  ) {
    const overlapDistance = previousColumnEvent.end - event.start;

    return baseEventPosition + overlapDistance + 10;
  }

  return baseEventPosition;
};

const findLatestEvent = events =>
  events.reduce(
    (currentLatest, event) => {
      if (currentLatest.end >= event.end) {
        return currentLatest;
      }

      return event;
    },
    { end: 0 }
  );

/**
 * Finds all overlapping events and groups them together in an array
 * @returns {Event[][]}
 */
export const getOverlappingEvents = events => {
  const sortedEvents = events.sort((a, b) => a.start - b.start);
  const eventGroups = [];

  sortedEvents.forEach(event => {
    if (eventGroups.length === 0) {
      eventGroups.push([event]);
    } else {
      // Since our events are sorted we always know the latest event group is the last one
      const latestEventGroup = eventGroups[eventGroups.length - 1];
      const earliestEventInGroup = latestEventGroup[0];
      const latestEventInGroup = findLatestEvent(latestEventGroup);

      if (
        event.start >= earliestEventInGroup.start &&
        event.start <= latestEventInGroup.end
      ) {
        latestEventGroup.push(event);
      } else {
        // no overlap
        eventGroups.push([event]);
      }
    }
  });

  return eventGroups;
};
