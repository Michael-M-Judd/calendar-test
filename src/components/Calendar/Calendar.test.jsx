import React from "react";
import { render } from "../../test-helpers";
import { Calendar } from "./Calendar";
import { testKey } from "./components/EventCard";

const events = [
  { start: 30, end: 150 },
  { start: 540, end: 600 },
  { start: 560, end: 620 },
  { start: 610, end: 670 }
];

describe("Calendar", () => {
  describe("with an array of events", () => {
    it("renders all the events", () => {
      const { getByTestId } = render(<Calendar initialEvents={events} />);

      events.forEach(event => {
        expect(getByTestId(testKey(event))).toBeInTheDocument();
      });
    });

    // I would usually use screenshot tests here rather than snapshots
    it("matches previous snapshot", () => {
      const { container } = render(<Calendar initialEvents={events} />);

      expect(container).toMatchSnapshot();
    });
  });
});
