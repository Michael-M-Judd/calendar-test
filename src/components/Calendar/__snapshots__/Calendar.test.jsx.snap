// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar with an array of events matches previous snapshot 1`] = `
<div>
  <div
    class="sc-fzXfLQ bXEwMe"
  >
    <section
      class="sc-AykKI kcbFWk"
    >
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            9:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            AM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          9:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            10:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            AM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          10:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            11:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            AM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          11:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            12:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            PM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          12:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            1:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            PM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          1:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            2:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            PM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          2:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            3:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            PM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          3:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            4:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            PM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          4:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            5:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            PM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          5:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            6:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            PM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          6:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            7:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            PM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          7:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            8:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            PM
          </span>
        </div>
        <div
          class="sc-AykKG inyJRW"
        >
          8:30
        </div>
      </div>
      <div
        class="sc-AykKH fRddoi"
      >
        <div>
          <span
            class="sc-AykKE evPEbk"
          >
            9:00
          </span>
          <span
            class="sc-AykKF iyMCzR"
          >
            PM
          </span>
        </div>
      </div>
    </section>
    <div
      class="sc-AykKC eEegzZ"
    >
      <div
        class="sc-AykKD enFPZo"
      >
        <div
          class="sc-fzXfLP jEQiaH"
          order="0"
        >
          <div
            class="sc-AykKJ kLaowu"
            data-testid="event-card:30:150"
            height="120"
          >
            <span
              class="sc-AykKK kcmvZX"
            >
              Sample Item
            </span>
            <span
              class="sc-fzXfLO iZACWc"
            >
              Sample Location
            </span>
          </div>
        </div>
        <div
          class="sc-fzXfLP ijIxhp"
          order="0"
        >
          <div
            class="sc-AykKJ lmblSj"
            data-testid="event-card:540:600"
            height="60"
          >
            <span
              class="sc-AykKK kcmvZX"
            >
              Sample Item
            </span>
            <span
              class="sc-fzXfLO iZACWc"
            >
              Sample Location
            </span>
          </div>
        </div>
        <div
          class="sc-fzXfLP hANstl"
          order="1"
        >
          <div
            class="sc-AykKJ lmblSj"
            data-testid="event-card:560:620"
            height="60"
          >
            <span
              class="sc-AykKK kcmvZX"
            >
              Sample Item
            </span>
            <span
              class="sc-fzXfLO iZACWc"
            >
              Sample Location
            </span>
          </div>
        </div>
        <div
          class="sc-fzXfLP dshTSj"
          order="2"
        >
          <div
            class="sc-AykKJ lmblSj"
            data-testid="event-card:610:670"
            height="60"
          >
            <span
              class="sc-AykKK kcmvZX"
            >
              Sample Item
            </span>
            <span
              class="sc-fzXfLO iZACWc"
            >
              Sample Location
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
